pkg/apis/storage/types.go

```diff
@@ -215,3 +215,174 @@ const (
 	// binding will occur during Pod scheduing.
 	VolumeBindingWaitForFirstConsumer VolumeBindingMode = "WaitForFirstConsumer"
 )
+
+// VolumePluginMechanism is the type of mechanism components should use for volume operations
+type VolumePluginMechanism string
+
+const (
+	// VolumePluginMechanismInTree means components should use the in-tree plugin for volume operations
+	VolumePluginMechanismInTree VolumePluginMechanism = "in-tree"
+	// VolumePluginMechanismCSI means components should use the CSI Driver for volume operations
+	VolumePluginMechanismCSI VolumePluginMechanism = "csi"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSIDriver captures information about a Container Storage Interface (CSI)
+// volume driver deployed on the cluster.
+// CSIDriver objects are non-namespaced.
+type CSIDriver struct {
+	metav1.TypeMeta
+
+	// Standard object metadata.
+	// metadata.Name indicates the name of the CSI driver that this object
+	// refers to; it MUST be the same name returned by the CSI GetPluginName()
+	// call for that driver.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ObjectMeta
+
+	// Specification of the CSI Driver.
+	Spec CSIDriverSpec
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSIDriverList is a collection of CSIDriver objects.
+type CSIDriverList struct {
+	metav1.TypeMeta
+
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// items is the list of CSIDriver
+	Items []CSIDriver
+}
+
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+	// attachRequired indicates this CSI volume driver requires an attach
+	// operation (because it implements the CSI ControllerPublishVolume()
+	// method), and that Kubernetes should call attach and wait for any attach
+	// operation to complete before proceeding to mounting.
+	// If value is not specified, default is false -- meaning attach will not be
+	// called.
+	// +optional
+	AttachRequired *bool
```

Comment:
As I understand in some cases a boolean pointer is used to ensure ambiguity associated with a boolean value which may not have been set in an existing type. But in this case since we are introducing a brand new type, do we still want to use boolean pointer?

Reply:
My understanding is that a boolean pointer is often used for optional parameters so we can differentiate the case when the parameter is not set vs the default value of false.

Reply:
But is there a difference between value not being set in JSON and value being false? I thought boolean pointers were more important for cases where a older client may not be aware of newly introduced field and hence not setting it and server might want to disambiguate between client not setting it vs defaulting to false. In this case - it doesn't appear to that case.

Reply:
I think it depends on what the caller is doing with this field.

Also since we are just moving from Alpha to Beta, should the APIs be kept the same (other than the API group/version) unless if there's something completely wrong?  What if we found out later on we need to differentiate between client not setting it vs default to false?

Reply:
I don't think we need to distinguish between true, false and unset for this particular case.  I don't think we absolutely need to keep the apis the same because it's a completely different api group.

Reply:
Removed pointer in the latest PR.

Reply:
There is logic missing in [defaults.go](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/pkg/apis/storage/v1/defaults.go) to default this field to true.

Reply:
Also @jsafrane do you see a need where we may need to extend this in the future to be a more general capabilities list?
ref https://github.com/kubernetes/kubernetes/pull/73883#discussion_r257396055